### ISSUE TITLE
Allow for base16 encoded binary fields

### ISSUE THREAD

**Issue by muelli** [elastic/elasticsearch#13836]
I've [searched for issues mentioning](https://github.com/elastic/elasticsearch/search?q=base16&type=Issues&utf8=%E2%9C%93) "base16" and didn't find any, so I'm daring to create a new issue.

I want to post binary data to ES, but that data is not base64 encoded. It is rather base16 encoded.

I tried to define a mapping with `{"type": "binary", "encoding":"base16"}` but that didn't work.

I'd appreciate if it was possible to define the encoding of the binary data posted.

**Comment by clintongormley** [comment]
Hi @muelli 

Yeah, this is the first request we've had for base 16.  I very much doubt we're going to support it, not least of all because of the 4x cost of storage.  

I'd build the conversion from base16->base64 into your indexing process.

**Comment by muelli** [comment]
Hah. I thought POSTing my data to ES was my indexing process ;-)

My naïve suggestion is to only treat the "encoding" property of the mapping's field as a hint to decode the data to then store in whatever way ES pleases. In a way to softcode the implicit base64 decoding of a binary field.

I don't have a need for the exact base16 data. It's just that I have a lot of JSON documents with data already base16 encoded. And I was hoping to have a more friendly way of getting my data into ES than  deserialising everything to JSON, decoding base16, encoding base64, and serialising everything back to JSON.

I know it's doable and I'm in the process of doing that. But maybe it makes a future me a tiny bit happier if I could load not only base64, but also base16, or pretty much anything into a binary field as long as there is an appropriate decoding function (think deflate, rot13, base2, ...).

**Comment by jpountz** [comment]
I'd rather settle on base64 and make sure this is what se use all the time to encode binary data.

**Comment by jprante** [comment]
@muelli you have the misconception that field type binary for Lucene can also drive the JSON string processing which happens before.

I have created a plugin that can process incoming base16 encoded JSON strings for buk indexing. The idea is to decode the base16 transparently in the XContent API and map them to a byte array that is passed on to the indexing where base64 is used internally.

See this plugin for an example 

https://github.com/jprante/elasticsearch-hex

Because the XContent API is not easy to extend when it comes to parse field strings, I implemented my own version by copy/paste, with a custom bulk action.

I think parsing incoming JSON strings for binary data encoding could take away some burden from users, not only for hexadecimal encodings, but also for processing "ASCII armored" data, like Ascii85/Base85, quoted-printable MIME strings, or percent encodings:

https://en.wikipedia.org/wiki/Binary-to-text_encoding
